File: ANS/examples/python/register_agent.py
"""
ANS Example: Register an agent with the ANS registry.

This shows how an agent registers itself, generating a keypair and
posting its capability manifest to an ANS registry node.

Prerequisites:
    pip install mcp-use[ans]  # (planned; not yet published)

Usage:
    python register_agent.py
"""

import asyncio
import json
from pathlib import Path

# These imports reflect the planned mcp_use.ans module structure.
# See ANS/ARCHITECTURE.md for the full SDK design.
# from mcp_use.ans import ANSClient, CapabilityManifest, KeyPair


# ─── Standalone demo (no actual mcp_use.ans yet) ────────────────────────────

import base64
import hashlib
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


@dataclass
class Capability:
    name: str
    description: str
    input_schema: dict[str, Any]
    output_schema: dict[str, Any] | None = None


@dataclass
class CapabilityManifest:
    ans_id: str
    name: str
    description: str
    version: str
    endpoint: str
    transports: list[str]
    capabilities: list[Capability]
    tags: list[str] = field(default_factory=list)
    rate_limit: str | None = None
    contact: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "ans_version": "0.1",
            "id": self.ans_id,
            "name": self.name,
            "description": self.description,
            "version": self.version,
            "endpoint": self.endpoint,
            "transport": self.transports,
            "capabilities": [
                {
                    "name": cap.name,
                    "description": cap.description,
                    "input_schema": cap.input_schema,
                    **({"output_schema": cap.output_schema} if cap.output_schema else {}),
                }
                for cap in self.capabilities
            ],
            "constraints": {
                **({"rate_limit": self.rate_limit} if self.rate_limit else {}),
            },
            "trust": {
                "public_key": "ed25519:PLACEHOLDER",  # Real impl: filled by KeyPair
                "signature": "PLACEHOLDER",  # Real impl: signed manifest hash
            },
            "meta": {
                "created_at": datetime.utcnow().isoformat() + "Z",
                "tags": self.tags,
                **({"contact": self.contact} if self.contact else {}),
            },
        }

    def canonical_json(self) -> bytes:
        """Returns the canonical JSON for signing (stable key ordering, no signature field)."""
        d = self.to_dict()
        d["trust"].pop("signature", None)
        return json.dumps(d, sort_keys=True, separators=(",", ":")).encode()

    def manifest_hash(self) -> str:
        return hashlib.sha256(self.canonical_json()).hexdigest()


# ─── Example usage ─────────────────────────────────────────────────────────

async def main():
    # 1. Define your agent's capability manifest
    manifest = CapabilityManifest(
        ans_id="search.my-org.web-researcher",
        name="Web Researcher",
        description=(
            "Searches the web and synthesizes information into structured reports. "
            "Supports English, Italian, and French queries."
        ),
        version="1.0.0",
        endpoint="https://agents.my-org.com/web-researcher",
        transports=["mcp", "http"],
        tags=["web", "search", "research"],
        rate_limit="100/hour",
        contact="platform@my-org.com",
        capabilities=[
            Capability(
                name="search",
                description="Search the web for information on a topic",
                input_schema={
                    "type": "object",
                    "required": ["query"],
                    "properties": {
                        "query": {
                            "type": "string",
                            "description": "The search query",
                        },
                        "max_results": {
                            "type": "integer",
                            "default": 10,
                            "minimum": 1,
                            "maximum": 50,
                        },
                        "language": {
                            "type": "string",
                            "enum": ["en", "it", "fr"],
                            "default": "en",
                        },
                    },
                },
                output_schema={
                    "type": "object",
                    "properties": {
                        "results": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "title": {"type": "string"},
                                    "url": {"type": "string"},
                                    "snippet": {"type": "string"},
                                },
                            },
                        },
                        "summary": {"type": "string"},
                    },
                },
            ),
        ],
    )

    # 2. Show the manifest that would be signed and submitted
    print("=== ANS Capability Manifest ===")
    print(json.dumps(manifest.to_dict(), indent=2))
    print()
    print(f"Manifest hash (for signing): {manifest.manifest_hash()}")
    print()

    # 3. When mcp_use.ans is available, registration would look like:
    print("=== Registration (pseudo-code, awaiting mcp_use.ans module) ===")
    print("""
    from mcp_use.ans import ANSClient, KeyPair

    # Generate or load your keypair
    keypair = KeyPair.load_or_generate("~/.ans/keys/web-researcher.pem")

    # Initialize the client
    client = ANSClient(registry="https://registry.ans.example.com")

    # Sign and register
    result = await client.register(manifest, keypair=keypair)
    print(f"Registered: {result.ans_id}")
    print(f"Certificate: {result.certificate}")
    print(f"Trust score: {result.trust_score}")
    """)

    # 4. Heartbeat (keep registration alive)
    print("=== Heartbeat (pseudo-code) ===")
    print("""
    import asyncio

    async def heartbeat_loop(client, ans_id, keypair):
        while True:
            await client.heartbeat(
                ans_id=ans_id,
                keypair=keypair,
                status="healthy",
                latency_p50_ms=120,
                latency_p99_ms=800,
            )
            await asyncio.sleep(60)  # Every minute
    """)


if __name__ == "__main__":
    asyncio.run(main())

File: ANS/examples/typescript/register-agent.ts
/**
 * ANS Example: Register an agent with the ANS registry (TypeScript).
 *
 * This shows how a TypeScript/Node.js agent registers itself, generating
 * a keypair and posting its capability manifest to an ANS registry node.
 *
 * Prerequisites:
 *   npm install mcp-use  # ans module planned in upcoming release
 *
 * Usage:
 *   npx tsx register-agent.ts
 */

import crypto from "node:crypto";

// ─── Types ─────────────────────────────────────────────────────────────────

interface Capability {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
}

interface CapabilityManifest {
  ans_version: string;
  id: string;
  name: string;
  description: string;
  version: string;
  endpoint: string;
  transport: string[];
  capabilities: Capability[];
  constraints?: {
    rate_limit?: string;
    max_context_tokens?: number;
    supported_languages?: string[];
    requires_auth?: boolean;
  };
  trust: {
    public_key: string;
    signature?: string;
  };
  meta: {
    created_at: string;
    updated_at?: string;
    tags?: string[];
    contact?: string;
    license?: string;
  };
}

// ─── Manifest Builder ──────────────────────────────────────────────────────

function buildManifest(): CapabilityManifest {
  return {
    ans_version: "0.1",
    id: "search.my-org.web-researcher",
    name: "Web Researcher",
    description:
      "Searches the web and synthesizes information into structured reports. " +
      "Supports English, Italian, and French queries.",
    version: "1.0.0",
    endpoint: "https://agents.my-org.com/web-researcher",
    transport: ["mcp", "http"],
    capabilities: [
      {
        name: "search",
        description: "Search the web for information on a topic",
        input_schema: {
          type: "object",
          required: ["query"],
          properties: {
            query: {
              type: "string",
              description: "The search query",
            },
            max_results: {
              type: "integer",
              default: 10,
              minimum: 1,
              maximum: 50,
            },
            language: {
              type: "string",
              enum: ["en", "it", "fr"],
              default: "en",
            },
          },
        },
        output_schema: {
          type: "object",
          properties: {
            results: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  title: { type: "string" },
                  url: { type: "string" },
                  snippet: { type: "string" },
                },
              },
            },
            summary: { type: "string" },
          },
        },
      },
    ],
    constraints: {
      rate_limit: "100/hour",
      max_context_tokens: 32000,
      supported_languages: ["en", "it", "fr"],
      requires_auth: false,
    },
    trust: {
      public_key: "ed25519:PLACEHOLDER", // Real impl: populated from KeyPair
    },
    meta: {
      created_at: new Date().toISOString(),
      tags: ["web", "search", "research"],
      contact: "platform@my-org.com",
      license: "proprietary",
    },
  };
}

// ─── Canonical JSON for signing ────────────────────────────────────────────

function canonicalJson(manifest: CapabilityManifest): Buffer {
  // Remove signature field before hashing, sort keys deterministically
  const { trust, ...rest } = manifest;
  const { signature: _, ...trustWithoutSig } = trust;
  const forSigning = { ...rest, trust: trustWithoutSig };
  return Buffer.from(JSON.stringify(forSigning, Object.keys(forSigning).sort()));
}

function manifestHash(manifest: CapabilityManifest): string {
  return crypto.createHash("sha256").update(canonicalJson(manifest)).digest("hex");
}

// ─── Main demo ─────────────────────────────────────────────────────────────

async function main() {
  const manifest = buildManifest();

  console.log("=== ANS Capability Manifest ===");
  console.log(JSON.stringify(manifest, null, 2));
  console.log();
  console.log(`Manifest hash (for signing): ${manifestHash(manifest)}`);
  console.log();

  // When mcp-use/ans is available, registration would look like:
  console.log("=== Registration (pseudo-code, awaiting mcp-use/ans module) ===");
  console.log(`
  import { ANSClient, KeyPair } from 'mcp-use/ans';

  // Generate or load keypair (stored in ~/.ans/keys/)
  const keypair = await KeyPair.loadOrGenerate('web-researcher');

  // Initialize the client
  const client = new ANSClient({ registry: 'https://registry.ans.example.com' });

  // Build manifest with real public key
  const manifest = buildManifest();
  manifest.trust.public_key = keypair.publicKeyEncoded;

  // Sign and register
  const result = await client.register(manifest, { keypair });
  console.log('Registered:', result.ansId);
  console.log('Certificate:', result.certificate);
  console.log('Trust score:', result.trustScore);
  `);

  // MCPServer auto-registration
  console.log("=== MCPServer Auto-Registration (pseudo-code) ===");
  console.log(`
  import { MCPServer } from 'mcp-use/server';
  import { ANSClient } from 'mcp-use/ans';

  const server = new MCPServer({
    name: 'web-researcher',
    version: '1.0.0',
    // ANS registration happens automatically on server.start()
    ans: {
      client: new ANSClient({ registry: 'https://registry.ans.example.com' }),
      ansId: 'search.my-org.web-researcher',
      privateKeyPath: '~/.ans/keys/web-researcher.pem',
      heartbeatIntervalMs: 60_000, // Report health every minute
    },
  });

  server.tool({
    name: 'search',
    description: 'Search the web',
    schema: z.object({ query: z.string() }),
    handler: async ({ query }) => {
      // ... implementation
    },
  });

  await server.start(); // Also registers to ANS
  `);
}

main().catch(console.error);
